Incorporate symboltoken to readers WIP

byronlin-bq opened the pull request:
Currently there are some integration tests failing, I am investigating


Issue #, if available:

Description of changes:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

matthewh-BQ commented on ion/binaryreader.go:
```diff
@@ -212,7 +213,12 @@ func (r *binaryReader) next() (bool, error) {
-			err := r.readLocalSymbolTable()
+
+			st, err := readLocalSymbolTable(r, r.cat)
+			if err == nil {
```
I might be misunderstanding something. This conditional is `true` when there is no error present? We'd then set `r.lst = st` and `return `false, nil`.

Should this conditional follow the same error handling pattern as above? `err != null`, then this conditional is `true` if an error is detected and then `return false, err`.? Otherwise, `r.lst = st` and `return true, nil`?

byronlin-bq replied:
> Otherwise, r.lst = st and return true, nil?

When we import a local symbol table, we will continue consuming the next data in the stream incase there are more local symbol tables to import, therefore we return `false,nil`

fernomac commented on ion/readerlocalsymboltable_test.go:
```diff
@@ -0,0 +1,189 @@
+)
+
+func TestLocalSymbolTableAppend(t *testing.T) {
+	text := "$ion_symbol_table::" +
```
Pro go tip: backtick-delimited ("raw") string literals are great for this kind of thing. See [here](https://github.com/amzn/ion-go/blob/master/ion/textwriter_test.go#L403) for an example.

byronlin-bq replied:
Thanks! , updated

fernomac commented on ion/binaryreader_test.go:
```diff
@@ -16,13 +16,171 @@
+const binaryFilePath = "testdata/binary/"
+
+func TestReadBinarySingleSymbol(t *testing.T) {
+	fileBytes := loadFile(t, binaryFilePath+"single_symbol.bindat")
```
`gofmt` has some [Opinions(tm)](https://github.com/golang/go/issues/12720) about this that have never made much sense to me, but are very solidly established.

(Any particular reason for loading this test data from a separate file instead of inlining it like the other tests here?)

byronlin-bq replied:
Removed test because it is covered by [TestBinaryReadSymbols](https://github.com/amzn/ion-go/blob/35232d755b0e08ebc28448694beec4801889c483/ion/binaryreader_test.go#L421)

fernomac commented on ion/binaryreader_test.go:
```diff
@@ -16,13 +16,171 @@
+	catalog := NewCatalog(table1, table2)
+
+	buf := bytes.Buffer{}
+	bw := NewBinaryWriter(&buf, table1, table2)
```
I generally tried to stay away from using the writers to test the readers and vice versa, lest there be a bug in both that round-trips data correctly but doesn't serialize it correctly per the spec. Could we generate this bit of test data once, make sure it looks correct, and then inline it?

byronlin-bq replied:
Removed this test and instead created `TestReadBinaryFieldNames` and `TestReadBinarySymbols` tests that takes in inline data

fernomac commented on ion/binaryreader_test.go:
```diff
@@ -16,13 +16,171 @@
+		t.Fatal("unable to step in")
+	}
+
+	r.Next()
+	currentType := r.Type()
+	if currentType != SymbolType {
+		t.Fatal("current type is not symbol.")
+	}
+	fns, err := r.FieldNameSymbol()
+	if err != nil {
+		t.Fatal("unable to get field name symbol")
+	}
+	if fns.Text != nil {
+		t.Fatal("field name symbol text is incorrect")
+	}
+	if fns.LocalSID != 10 {
+		t.Fatal("field name symbol sid is incorrect")
+	}
```
Creating versions of `_symbolAF` and friends that deal in SymbolTokens might make the intent of these sections more clear.

byronlin-bq replied:
Modified _symbolAF to support SymbolToken validation. Also added _nextF to support FieldNameSymbol validation

fernomac commented on ion/textreader.go:
```diff
@@ -285,6 +294,16 @@ func (t *textReader) nextBeforeTypeAnnotations() (bool, error) {
 		t.valueType = StructType
 		t.value = StructType
+
+		if isIonSymbolTable(t.Annotations()) {
```
Should we also check if we're at the top level?

byronlin-bq replied:
Added top level check

fernomac commented on ion/textreader.go:
```diff
@@ -389,6 +408,23 @@ func (t *textReader) verifyUnquotedSymbol(val string, ctx string) error {
 }
 
+// IsSid checks if the text is a symbol identifier.
+func isSid(text string) bool {
```
Can you use `isSymbolRef` here?

byronlin-bq replied:
Updated

byronlin-bq commented on ion/reader.go:
```diff
@@ -163,6 +163,17 @@ type Reader interface {
+	SymbolValue() (SymbolToken, error)
+
+	// Clear clears the current value from the reader.
+	clear()
```
I tried removing clear() and [TestReadNullLST](https://github.com/amzn/ion-go/blob/master/ion/binaryreader_test.go#L60-L69) test failed. This is because it is checking for [empty annotations.](https://github.com/amzn/ion-go/blob/master/ion/textreader_test.go#L820-L822)

I updated clear to be exportable

zslayton replied:
I don't think we want `Clear()` to be part of the public API that users can implement. It feels like an implementation detail to me; no user would ever call `Clear()`.

In OO languages, you might have a `Writer` interface and an `AbstractWriter` base class to hold common implementation details like `Clear()`. I'm unsure what the best way to model that in Go would be.

byronlin-bq replied:
Removed `Clear()` from `Reader` interface. `Clear()` was being used in the beginning of [readLocalSymbolTable()](https://github.com/amzn/ion-go/pull/149/commits/851ab03c6a785fb196e6722d01ed0be28ed4ae9f#diff-c7a62f6b5becd9a9a65eb57ed6c40180L7-L10) and I move that logic to to text/binary readers just before they call `readLocalSymbolTable`

byronlin-bq commented on ion/binaryreader.go:
```diff
@@ -153,7 +154,7 @@ func (r *binaryReader) next() (bool, error) {
 				return false, err
 			}
-			r.value = r.resolve(id)
+			r.value = id
```
Updated

byronlin-bq commented on ion/textreader_test.go:
```diff
@@ -844,3 +866,37 @@ func _eof(t *testing.T, r Reader) {
 	}
 }
+
+func assertZeroSymbolTest(t *testing.T, r Reader) {
```
Removed for now

byronlin-bq commented on ion/binaryreader_test.go:
```diff
@@ -16,13 +16,171 @@
+	r.Next()
+	currentType := r.Type()
+	if currentType != SymbolType {
+		t.Fatal("current type is not symbol.")
```
Removed

byronlin-bq commented on ion/binaryreader_test.go:
```diff
@@ -16,13 +16,171 @@
+	r.Next()
+	currentType = r.Type()
+	if currentType != SymbolType {
+		t.Fatal("current type is not symbol.")
```
Removed

byronlin-bq commented on ion/binaryreader_test.go:
```diff
@@ -16,13 +16,171 @@
+		t.Fatal("symbol token sid is incorrect")
+	}
+
+	//make sure that a reader with the correct imports can read it
```
Updated

byronlin-bq commented on ion/readerlocalsymboltable_test.go:
```diff
@@ -0,0 +1,189 @@
+	systemTable := imports[0]
+	systemMaxID := systemTable.MaxID()
+
+	checkSymbol(t, "s1", systemMaxID+1, st)
```
gofmt removes the spacing between appending statements

byronlin-bq commented on ion/textreader.go:
```diff
@@ -389,6 +408,23 @@ func (t *textReader) verifyUnquotedSymbol(val string, ctx string) error {
+	if string(text[0]) == "$" {
+		if len(text) > 1 {
+			for i := 1; i < len(text); i++ {
+				// check if value is a digit.
```
Updated

byronlin-bq commented on ion/binaryreader.go:
```diff
@@ -501,3 +336,53 @@ func (r *binaryReader) StepOut() error {
+		return "", nil
+	}
+
+	// check if value is symbol(in the form of SID) or string.
```
Updated

byronlin-bq commented on ion/reader.go:
```diff
@@ -199,10 +210,11 @@ type reader struct {
-	valueType   Type
-	value       interface{}
+	fieldName    *string
+	fieldnameSID int64
```
Updated in this [commit](https://github.com/amzn/ion-go/pull/149/commits/9a87cc8c89c1a04102c2d577c06c32e5c191ae8a)

byronlin-bq commented on ion/reader.go:
```diff
@@ -372,17 +384,6 @@ func (r *reader) TimestampValue() (Timestamp, error) {
 }
 
-// StringValue returns the current value as a string.
-func (r *reader) StringValue() (string, error) {
```
Updated

byronlin-bq commented:
> gofmt has some Opinions(tm) about this that have never made much sense to me, but are very solidly established.

(Any particular reason for loading this test data from a separate file instead of inlining it like the other tests here?)

Removed and added test that has inline ion data

justing-bq commented on ion/writer.go:
```diff
@@ -164,6 +168,24 @@ func (w *writer) FieldName(val string) error {
+// It may only be called while writing a struct.
+func (w *writer) FieldNameSymbol(val SymbolToken) error {
+	if !w.IsInStruct() {
+		w.err = errors.New("ion: Writer.FieldNameSymbol called when not writing a struct")
```
errors.New? 
Is there a suitable error type in err.go we can use instead? If not can we create a new error type in err.go?

byronlin-bq replied:
Updated to UsageError

byronlin-bq commented on ion/reader.go:
```diff
@@ -163,6 +163,17 @@ type Reader interface {
 	// IsInStruct indicates if the reader is currently positioned in a struct.
 	IsInStruct() bool
+
+	// FieldNameSymbol returns the field name associated with the current value as a SymbolToken.
+	// It returns an error if the current value has no field name.
+	FieldNameSymbol() (SymbolToken, error)
+
+	// SymbolValue returns the SymbolToken associated with the current value. It returns an
+	// error if the current value is not an Ion symbol.
+	SymbolValue() (SymbolToken, error)
```
Would like to get opinions from @therapon, @zslayton

therapon replied:
We made a similar decision before. I vote for moving these to `*SymbolToken`

byronlin-bq replied:
Updated to return `(*SymbolToken, error)`

zslayton commented on ion/textreader.go:
```diff
@@ -183,6 +187,16 @@ func (t *textReader) nextBeforeFieldName() (bool, error) {
 		}
 
+		if tok == tokenSymbolQuoted {
+			t.fieldNameSymbol = SymbolToken{Text: &val, LocalSID: SymbolIDUnknown}
```
It's hard to tell at a glance without an IDE: is the value of `SymbolIDUnknown` zero?

byronlin-bq replied:
`SymbolIDUnknown` has value `-1`

zslayton commented on ion/textreader.go:
```diff
@@ -425,6 +459,31 @@ func (t *textReader) onSymbol(val string, tok token, ws bool) error {
 	return nil
 }
 
+// readSymbol reads a text and returns a symbol token.
```
Could you flesh out this comment a bit?

therapon replied:
From the code this looks like a function that can take a SID as a Go `string` and use it as a SID, i.e., `readSymbol("$42")`. If that is indeed the intent, maybe call this `readAsSID`?

byronlin-bq replied:
> From the code this looks like a function that can take a SID as a Go string and use it as a SID, i.e., readSymbol("$42"). If that is indeed the intent, maybe call this readAsSID?

The function reads SID text as well as symbol text e.g., `foo` and return a symboltoken.

Updated comment to:
```
// readSymbol reads a text as either a symbol identifier e.g., "$42" or a symbol text e.g., "foo" and resolves to a
// symbol token.
```

zslayton commented on ion/textutils.go:
```diff
@@ -319,3 +319,7 @@ func parseInt(str string, radix int) (interface{}, error) {
 }
+
+func isStringValueNil(errMsg string) bool {
+	return errMsg == "ion: usage error in Reader.StringValue: string value is unknown"
```
I think this is a strong sign that we have an API that should be returning `*string` instead of `string`. Requiring users to do a string comparison against an error message that could change without notice is very brittle.

byronlin-bq replied:
Updated stringvalue() to return `(*string, error)`

therapon commented on ion/binaryreader.go:
```diff
@@ -248,192 +259,34 @@ func (r *binaryReader) readBVM() error {
+// ReadSymbol reads an ID and returns a symbol token.
+func (r *binaryReader) readSymbol(id uint64) (SymbolToken, error) {
+	if id > r.SymbolTable().MaxID() {
+		return symbolTokenUndefined, &UsageError{"Reader.Next", "sid is out of range "}
```
I would add the value for `id` and `MaxID()` in the error message

byronlin-bq replied:
Updated

fernomac commented on ion/binaryreader_test.go:
```diff
@@ -401,6 +381,36 @@ func TestReadBinaryFloats(t *testing.T) {
+		0x71, 0x6F, // $111
+	})
+
+	_symbolAF(t, r, nil, nil, "", SymbolToken{Text: nil, LocalSID: 11}, true, false, false)
```
The AF suffix means this function accepts the expected [a]nnotations and [f]ield name as parameters, as opposed to _symbol which asserts that the value has neither. The slang implication was a happy coincidence. :) It should probably be changed or at least documented.

byronlin-bq replied:
Added docstring to `_symbolAF` function

byronlin-bq commented on ion/binaryreader_test.go:
```diff
@@ -301,25 +268,38 @@ func TestReadBinaryStrings(t *testing.T) {
+	r.Next()
+	r.StepIn()
+	_nextF(t, r, nil, SymbolToken{Text: nil, LocalSID: 0}, false, false)
+	_nextF(t, r, newString("$ion"), SymbolToken{Text: newString("$ion"), LocalSID: 1}, false, false)
```
It came from this [issue](https://github.com/amzn/ion-go/issues/11) and [PR](https://github.com/amzn/ion-go/pull/14)

byronlin-bq commented on ion/binaryreader.go:
```diff
@@ -248,192 +259,34 @@ func (r *binaryReader) readBVM() error {
-	}
-
-	if err := r.StepOut(); err != nil {
+	st, err := r.readSymbol(id)
```
Updated to `resolveSymbolID `

byronlin-bq commented on ion/reader.go:
```diff
@@ -406,3 +406,51 @@ func (r *reader) clear() {
+	}
+
+	if r.valueType != StringType && r.valueType != SymbolType {
+		return "", &UsageError{"Reader.StringValue", "value is not a string"}
```
Updated

byronlin-bq commented on ion/reader.go:
```diff
@@ -406,3 +406,51 @@ func (r *reader) clear() {
+	}
+
+	if r.value == nil {
+		return "", &UsageError{"Reader.StringValue", "string value is unknown"}
```
Updated stringValue() to return `(*string, error)`

byronlin-bq commented on ion/reader.go:
```diff
@@ -406,3 +406,51 @@ func (r *reader) clear() {
+	}
+
+	if st.Text == nil {
+		return "", &UsageError{"Reader.StringValue", "string value is unknown"}
```
No-op because I updated stringValue() to return `(*string, error)`

byronlin-bq commented on cmd/ion-go/eventwriter.go:
```diff
@@ -41,6 +41,11 @@ func NewEventWriter(out io.Writer) ion.Writer {
 }
 
+func (e *eventwriter) FieldNameSymbol(val ion.SymbolToken) error {
+	e.fieldname = val.Text
```
I added `FieldNameSymbol()` to the writer interface so I can test the integration tests.

Updated to:
```func (e *eventwriter) FieldNameSymbol(val ion.SymbolToken) error {
	if val.Text == nil {
		sid := fmt.Sprintf("$%v", val.LocalSID)
		e.fieldname = &sid
		return nil
	}

	e.fieldname = val.Text
	return nil
}```

zslayton commented on ion/unmarshal_test.go:
```diff
@@ -340,11 +340,11 @@ func TestUnmarshalBinary(t *testing.T) {
 	var stringVal string
 	stringBytes := prefixIVM([]byte{0x83, 'a', 'b', 'c'}) // "abc"
-	test(stringBytes, stringVal, "abc")
+	test(stringBytes, stringVal, newString("abc"))
```
I'm not sure I understand the value of [the `newString` function](https://github.com/amzn/ion-go/blob/58d1539a051c268383698a3566e3ef176d994c1e/ion/symboltoken_test.go#L25-L27):

```go
func newString(value string) *string {
	return &value
}
```

It's certainly correct, but why is `newString("hello")` preferable to `&"hello"`?

fernomac replied:
Because only one of them successfully compiles? :) It’s a bit of a chicken-and-egg: it’s not very “go-ey” to use pointers to basic types in APIs, so there isn’t a lot of built-in syntax sugar to make working with them easy (in this case: you can’t directly take the address of a literal, you need to assign it to a variable and then take the address of the variable), which reinforces the convention against using them in APIs.

zslayton replied:
> Because only one of them successfully compiles? :)

That's a good reason. :)

> you can’t directly take the address of a literal, you need to assign it to a variable and then take the address of the variable

Huh.

justing-bq replied:
This isn't a choice between `newString("hello")` and `&"hello"`. 
Rather it's a choice between `newString("hello")` and 
```go
str := "hello"
&str
```

zslayton replied:
I understood what @fernomac was describing, I was just surprised to learn that that was a restriction in Go.

zslayton commented on ion/integration_test.go:
```diff
@@ -689,8 +680,9 @@ func readCurrentValue(t *testing.T, reader Reader) ionItem {
 	currentType := reader.Type()
-	if reader.IsNull() {
-		ionItem.value = append(ionItem.value, textNulls[currentType])
+	// Excluding SymbolType from setting value to 'null.symbol' because readers can read "'null.symbol'" as text: 'null.symbol'.
```
> because readers can read "'null.symbol'" as text: 'null.symbol'

I don't think I understand. If a text reader encounters a `null.symbol`, it has an ion type of `symbol`, `IsNull()` is `true`, and `StringValue()` would return `nil`. 

How would a reader read it as text?

byronlin-bq replied:
This stems from this [test](https://github.com/amzn/ion-tests/blob/master/iontestdata/good/non-equivs/symbols.ion#L25-L26).

According to the symbol specs, if a text reader encounters a single quoted text, it has an ion type of symbol, IsNull() is false, and StringValue() would return the text.

In this case, StringValue() will return `null.symbol` which is the same value appended by the test when text reader encounters unquoted text `null.symbol`

zslayton replied:
> According to the symbol specs, if a text reader encounters a single quoted text, it has an ion type of symbol, IsNull() is false, and StringValue() would return the text.

Agreed. I think my confusion stems from the comment's phrasing:

> because readers **can read** "'null.symbol'" as text: 'null.symbol'.

If a `null.symbol` is wrapped in single quotes (`'null.symbol'`), it *must* be read as text, and never as a null.

> In this case, StringValue() will return "null.symbol" which is the same value appended by the test when text reader encounters unquoted text null.symbol

I would expect `reader.StringValue()` to return:
* `"null.symbol"` when the Ion value is `'null.symbol'` (a symbol)
* An `err` (or `nil`, depending on #152) when the Ion value is `null.symbol` (a null)

Is that what you're describing, or am I missing something?

byronlin-bq replied:
> Is that what you're describing, or am I missing something?

That is correct

zslayton commented on ion/reader.go:
```diff
@@ -155,14 +155,25 @@ type Reader interface {
 	// StringValue returns the current value as a string (if that makes sense). It returns
 	// an error if the current value is not an Ion symbol or an Ion string.
-	StringValue() (string, error)
+	StringValue() (*string, error)
```
This really simplifies [the case from my earlier comment](https://github.com/amzn/ion-go/pull/149#discussion_r481272560); if the string or symbol was `null`, this function will return a `nil` that can be passed around. Great.

However, we now have a larger consistency question for the `Reader` interface. Should all values return `*type` so they return a `nil` if the Ion value is `null`? At the moment most of them return an `err` with the expectation that you will have handled `null` separately through a combination of `IsNull()` and `Type()`. (This isn't documented yet, by the way. See [`IntValue()` for an example](https://github.com/amzn/ion-go/blob/1f421dfa4f274ec841989150ea325434170aac0f/ion/reader.go#L125-L128) -- the doc comment doesn't explain what to expect if the value is `null`.)

Having two layers of checking to do after calling each `*Value()` function (Is it an `err`? Is it `nil`?) feels pretty heavy to me, which is why I think we were generating a `UsageError` for `null` values originally.

This is a larger discussion, but it's something I think we need to get right before 1.0. I'm ok with opening an issue to track this so it doesn't block the PR.

byronlin-bq replied:
Created ticket: https://github.com/amzn/ion-go/issues/152

justing-bq commented on ion/readlocalsymboltable.go:
```diff
@@ -0,0 +1,206 @@
+
+	for r.Next() {
+		var err error
+		switch *r.FieldName() {
```
What if r.FieldName() is nil? Let's add an error case for that.

byronlin-bq replied:
Added a check for nil fieldname
